Feat: add contribution guidelines (#5)
TheAlgorithms/TypeScript: CONTRIBUTING.md

# Contributing guidelines

## Before contributing

Welcome to [TheAlgorithms/TypeScript](https://github.com/TheAlgorithms/TypeScript)! Before sending your pull requests,
make sure that you **read the whole guidelines**. If you have any doubt on the contributing guide, please feel free to
[state it clearly in an issue](https://github.com/TheAlgorithms/TypeScript/issues/new).

## Contributing

### Contributor

We are very happy that you consider implementing algorithms and data structures for others! This repository is
referenced and used by learners from around the globe. Being one of our contributors, you agree and confirm that:

- You did your work - plagiarism is not allowed.
  - Any plagiarized work will not be merged.
- Your work will be distributed under [MIT License](LICENSE) once your pull request is merged.
- Your submitted work must fulfill our styles and standards.

**New implementation** is welcome! For example, new solutions to a problem, different representations of a graph data
structure or algorithm designs with different complexity.

> **Improving comments** and **writing proper tests** are also highly welcome.

Reviewer: Which framework is used for testing?

Author: For now, none. I think Jest is going to be added in the future.

#### File Naming Convention

- filenames should use the UpperCamelCase (PascalCase) style.
- There should be no spaces in filenames.
- **Example:** `UserProfile.ts` is allowed but `userprofile.ts`,`Userprofile.ts`,`user-Profile.ts`,`userProfile.ts` are
  not.

#### Module System

We use the [ES Module](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/) system, which bring an official, standardized module system to JavaScript.

It roughly means you will need to use `export` and `import` statements instead of `module.exports` and `require()`.

#### Coding Style

To maximize the readability and correctness of our code, we require that new submissions follow the
[TypeScript Standard Style](https://github.com/standard/ts-standard).

In order to apply the coding style (where it can be done automatically). If an error is shown, please figure out what's
wrong, fix it and run standard again.

A few (but not all) of the things to keep in mind:

- Use camelCase with the leading character as lowercase for identifier names (variables and functions).
- Names start with a letter.
- Follow code indentation: Always use 2 spaces for indentation of code blocks.

```ts

> function sumOfArray(arrayOfNumbers: number[]): number {

Reviewer: Much redundant naming: Types are made part of the name here. IMO the function should be just `sum(arr)` or `sum(nums)` or the like - the rest is clear from the type annotations.

# Contributing guidelines

## Before contributing

Welcome to [TheAlgorithms/TypeScript](https://github.com/TheAlgorithms/TypeScript)! Before sending your pull requests,
make sure that you **read the whole guidelines**. If you have any doubt on the contributing guide, please feel free to

> [state it clearly in an issue](https://github.com/TheAlgorithms/TypeScript/issues/new).

Reviewer 2: ```suggestion
[state it clearly in an issue](https://github.com/TheAlgorithms/TypeScript/issues/new) or on our [**Discord server**](https://discord.gg/c7MnfGFGa6).
```

## Contributing

### Contributor

We are very happy that you consider implementing algorithms and data structures for others! This repository is
referenced and used by learners from around the globe. Being one of our contributors, you agree and confirm that:

- You did your work - plagiarism is not allowed.
  - Any plagiarized work will not be merged.
- Your work will be distributed under [MIT License](LICENSE) once your pull request is merged.
- Your submitted work must fulfill our styles and standards.

> **New implementation** is welcome! For example, new solutions to a problem, different representations of a graph data

Reviewer 2: ```suggestion
**New implementations** are welcome! For example, new solutions to a problem, different representations of a graph data
```

> structure or algorithm designs with different complexity.

Reviewer 2: ```suggestion
structure, or algorithm designs with different complexity.
```

**Improving comments** and **writing proper tests** are also highly welcome.

### Contribution

We appreciate any contribution, from fixing grammar mistakes to implementing complex algorithms. Please read this
section if you are contributing to your work.

If you submit a pull request that resolves an open issue, please help us to keep our issue list small by adding

> `fixes: #{$ISSUE_NO}` to your commit message. GitHub will use this tag to auto-close the issue if your PR is merged.

Reviewer 2: `fixes` is usually for closing issues outside the repository.
```suggestion
`closes: #{$ISSUE_NO}` to your commit message. GitHub will use this tag to auto-close the issue if your PR is merged.
```

#### What is an Algorithm?

An Algorithm is one or more functions (or classes) that:

- take one or more inputs.
- perform some internal calculations or data manipulations.
- return one or more outputs.
- have minimal side effects.

Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.

Algorithms should:

- have intuitive class and function names that make their purpose clear to readers.
- use TypScript naming conventions and intuitive variable names to ease comprehension.
- be flexible to take different input values.
- raise TypeScript exceptions (RangeError, etc.) on erroneous input values.

Algorithms in this repo should not be how-to examples for existing TypeScript packages. Instead, they should perform
internal calculations or manipulations to convert input values into different output values. Those calculations or
manipulations can use data types, classes, or functions of existing TypeScript packages but each algorithm in this repo
should add unique value.

#### File Naming Convention

> - filenames should use the UpperCamelCase (PascalCase) style.

Reviewer 2: ```suggestion
- Filenames should use the UpperCamelCase (PascalCase) style.
```

- There should be no spaces in filenames.

> - **Example:** `UserProfile.ts` is allowed but `userprofile.ts`,`Userprofile.ts`,`user-Profile.ts`,`userProfile.ts` are
>   not.

Reviewer 2: ```suggestion
- **Example:** `UserProfile.ts` is allowed. Do not use `userprofile.ts`, `Userprofile.ts`, `user-Profile.ts`, `userProfile.ts`; these methods are discouraged.
```

> 

Reviewer 2: Should we include the use of semantic commit messages? These make it much more professional, IMO.

Reviewer 3: I usually use the following format "<samp>commit type: short description</samp>" and put any extra information in the description. Some examples:
<samp>
feat: release candidate for v0.1.2
chore: fix spelling
fix: improper error message
</samp>

#### Commit Messages Formatting

- Prefer to use the following format: `<type>: <short summary>`. If necessary, put any extra information in the description.
- Commit types include (but are not limited to): 
  - **docs**: Documentation only changes
  - **feat**: A new feature
  - **fix**: A bug fix
  - **test**: Adding or fixing tests
  - **chore**: CI / code quality / minor quality of life improvements

- **Examples**: 
  - `feat: add quicksort algorithm`
  - `chore: fix spelling`
  - `fix: improper error message`
  - `docs: add contributing guidelines`
  - `test: add test for quicksort algorithm`

#### Module System

We use the [ES Module](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/) system, which brings an official, standardized module system to TypeScript.

It roughly means you will need to use `export` and `import` statements instead of `module.exports` and `require()`.

#### Coding Style

To maximize the readability and correctness of our code, we require that new submissions follow the
[TypeScript Standard Style](https://github.com/standard/ts-standard).

A few (but not all) of the things to keep in mind:

> - Use `lowerCamelCase` with the leading character as lowercase for identifier names (variables and functions).

Reviewer 3: Add a point to use `UpperCamelCase` for class, type, and interface names.
And to use `PASCAL_CASE` for universal constants.

- Use camelCase with the leading character as lowercase for identifier names (variables and functions).
- Names start with a letter.
- Follow code indentation: Always use 2 spaces for indentation of code blocks.

```ts
function sum(arr: number[]): number {
  let total = 0;
  for (let i = 0; i < arr.length; i++) {
    total += arr[i];
  }
  return total;
}
```

- Avoid using global variables and avoid `==`.
- Please use `let` over `var`.
- Please refrain from using `console.log` or any other console methods.
- **Absolutely** don't use `alert`.
- We strongly recommend the use of ECMAScript 6.
- Avoid the use of `any` type. Create an interface/class instead.
- Don't prefix Interfaces with `I`.
- Prefer not to use `null` nor `undefined` for explicit unavailability. 

```ts
// BAD
let foo = { x: 123, y: undefined };

// GOOD
let foo: { x: number, y?: number } = { x:123 };
```
- Annotate arrays as `foos: Foo[]` instead of `foos: Array<Foo>`.
- Avoid importing external libraries for basic algorithms. Only use those libraries for complicated algorithms.
- Most importantly:
  - **Be consistent in the use of these guidelines when submitting.**
  - Happy coding!

> Writer [@gefgu](https://github.com/gefgu), September 2022.

Reviewer 2: We should include the community as well. 🙂
```suggestion
Written by [**@gefgu**](https://github.com/gefgu) and community, September 2022.
```

make sure that you **read the whole guidelines**. If you have any doubt on the contributing guide, please feel free to
[state it clearly in an issue](https://github.com/TheAlgorithms/TypeScript/issues/new) or on our [**Discord server**](https://discord.gg/c7MnfGFGa6).

## Contributing

### Contributor

We are very happy that you consider implementing algorithms and data structures for others! This repository is
referenced and used by learners from around the globe. Being one of our contributors, you agree and confirm that:

- You did your work - plagiarism is not allowed.
  - Any plagiarized work will not be merged.
- Your work will be distributed under [MIT License](LICENSE) once your pull request is merged.
- Your submitted work must fulfill our styles and standards.

**New implementations** are welcome! For example, new solutions to a problem, different representations of a graph data
structure, or algorithm designs with different complexity.

**Improving comments** and **writing proper tests** are also highly welcome.

### Contribution

We appreciate any contribution, from fixing grammar mistakes to implementing complex algorithms. Please read this

> section if you are contributing to your work.

Reviewer: "if you are contributing to your work"?

If you submit a pull request that resolves an open issue, please help us to keep our issue list small by adding
`closes: #{$ISSUE_NO}` to your commit message. GitHub will use this tag to auto-close the issue if your PR is merged.

#### What is an Algorithm?

An Algorithm is one or more functions (or classes) that:

- take one or more inputs.
- perform some internal calculations or data manipulations.
- return one or more outputs.
- have minimal side effects.

Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.

Algorithms should:

- have intuitive class and function names that make their purpose clear to readers.

> - use TypScript naming conventions and intuitive variable names to ease comprehension.

Reviewer: TypeScript*

- be flexible to take different input values.
- raise TypeScript exceptions (RangeError, etc.) on erroneous input values.

Algorithms in this repo should not be how-to examples for existing TypeScript packages. Instead, they should perform
internal calculations or manipulations to convert input values into different output values. Those calculations or
manipulations can use data types, classes, or functions of existing TypeScript packages but each algorithm in this repo
should add unique value.

#### File Naming Convention

- Filenames should use the UpperCamelCase (PascalCase) style.
- There should be no spaces in filenames.

> - **Example:** `UserProfile.ts` is allowed. Do not use `userprofile.ts`, `Userprofile.ts`, `user-Profile.ts`, `userProfile.ts`; these methods are discouraged.

Reviewer: "methods"? naming conventions

#### Commit Messages Formatting

- Prefer to use the following format: `<type>: <short summary>`. If necessary, put any extra information in the description.
- Commit types include (but are not limited to): 

>   - **docs**: Documentantion only changes

Reviewer: Documentation*

  - **feat**: A new feature
  - **fix**: A bug fix

>   - **chore**: Miscellaneous stuff that does not match any of the above

Reviewer: I wouldn't use chore for "anything else", especially as this doesn't cover tests yet; would adding tests be a "feat"? "chore" is for CI / code quality / minor quality of life improvements like fixing typos etc. which doesn't impact the way the code works or is tested IMO.

- **Examples**: 
  - `feat: add quicksort algorithm`
  - `chore: fix spelling`
  - `fix: improper error message`
  -  `docs: add contributing guidelines`

#### Module System

> We use the [ES Module](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/) system, which bring an official, standardized module system to JavaScript.

Reviewer: brings*; also JavaScript?

It roughly means you will need to use `export` and `import` statements instead of `module.exports` and `require()`.

#### Coding Style

To maximize the readability and correctness of our code, we require that new submissions follow the
[TypeScript Standard Style](https://github.com/standard/ts-standard).

> In order to apply the coding style (where it can be done automatically). If an error is shown, please figure out what's

Reviewer: The sentence
> In order to apply the coding style (where it can be done automatically).

is not complete

wrong, fix it and run standard again.

A few (but not all) of the things to keep in mind:

> - Use camelCase with the leading character as lowercase for identifier names (variables and functions).

Reviewer: `lowerCamelCase`

- Names start with a letter.
- Follow code indentation: Always use 2 spaces for indentation of code blocks.

```ts
function sum(arr: number[]): number {
  let total = 0;

>   for (let i = 0; i < arr.length; i++) {

Reviewer: Just saying: Wouldn't `for (const elem of arr) total += elem` be most idiomatic here (apart from using `arr.reduce`)?

    total += arr[i];
  }
  return total;
}
```

> - Avoid using global variables and avoid `==`.

Reviewer: (use `===` instead)

> - Please use `let` over `var`.

Reviewer: Use only `let` and `const`, never use `var`.

> - Please refrain from using `console.log` or any other console methods.

Reviewer: This and the following point about `alert` is just an instance of "prefer proper input/output of your functions over side effects"?

- **Absolutely** don't use `alert`.

> - We strongly recommend the use of ECMAScript 6.

Reviewer: Only "strongly recommend"? I'd say we *require* it (also ECMAScript? This is TS).

> - Avoid the use of `any` type. Create an interface/class instead.

Reviewer: (create proper types instead)

Should be something like "only use `any` if appropriate". I can imagine f.E. k/v data structures using `any` for their values: You don't care which type the values are; you don't have to apply any operation to them, you just pass them around and store them.

> - Don't prefix Interfaces with `I`.

Reviewer: interfaces* and this should perhaps be a more general rule (no redundant naming: don't prefix class members with `m`, functions with `func` or `f` etc.)

> - Prefer not to use `null` nor `undefined` for explicit unavailability. 

Reviewer: ``Prefer using optional fields over `null` or `undefined` ``

```ts
// BAD
let foo = { x: 123, y: undefined };

// GOOD
let foo: { x: number, y?: number } = { x:123 };
```
- Annotate arrays as `foos: Foo[]` instead of `foos: Array<Foo>`.

> - Avoid importing external libraries for basic algorithms. Only use those libraries for complicated algorithms.

Reviewer: I'm leaning towards forbidding external libraries in the implementations of algorithms entirely (except for our tooling such as tests), requiring implementations to be "from scratch".

- Most importantly:
  - **Be consistent in the use of these guidelines when submitting.**
  - Happy coding!

> Written by [**@gefgu**](https://github.com/gefgu) and community, September 2022.

Reviewer: This will be visible in the Git history. Is it necessary / beneficial to put the main author & date of writing here?
